metric_learn/constraints.py

```diff
@@ -33,6 +35,56 @@ def positive_negative_pairs(self, num_constraints, same_length=False,
       return a[:n], b[:n], c[:n], d[:n]
     return a, b, c, d
 
+  def generate_knntriplets(self, X, k_genuine, k_impostor):
+
+    labels = np.unique(self.partial_labels)
+    L = len(labels)
+    len_input = np.size(self.partial_labels, 0)
+    triplets = np.empty((len_input*k_genuine*k_impostor, 3), dtype=np.intp)
+
+    start = 0
+    finish = 0
+    neigh = NearestNeighbors()
+
+    for i in range(L):
+
+        # generate mask for current label
+        gen_mask = self.partial_labels == labels[i]
+        gen_indx = np.where(gen_mask)
+
+        # get k_genuine genuine neighbours
+        neigh.fit(X=X[gen_indx])
+        gen_neigh = np.take(gen_indx, neigh.kneighbors(n_neighbors=k_genuine,
```

Comment:
I think here you are including the point itself as its own neighbor

Reply:
No! The same point is not considered when `X` is not provided, as stated in `sklearn.neighbors.NearestNeighbors` documentation:
> (...) If not provided, neighbors of each indexed point are returned. In this case, the query point is not considered its own neighbor.

Reply:
OK, my bad!